url_matcher/matcher.py

```diff
@@ -151,14 +150,22 @@ def remove(self, identifier: Any):
     def get(self, identifier: Any) -> Optional[Patterns]:
         return self.patterns.get(identifier)
 
-    def match(self, url: str) -> Optional[Any]:
-        return next(self.match_all(url), None)
+    def match(self, url: str, *, include_universal=True) -> Optional[Any]:
+        return next(self.match_all(url, include_universal=include_universal), None)
 
-    def match_all(self, url: str) -> Iterator[Any]:
+    def match_all(self, url: str, *, include_universal=True) -> Iterator[Any]:
```

Comment:
I can't make a good short name for this, as it's not strictly about including/excluding or about preferring, we always return non-universal matches and then, depending on the flag, either always return universal ones or do that only if there were no non-universal matches.

Reply:
As a user, this is a bit confusing to me since I've set `include_universal=False` but an ID was still returned to me. For example, this test taken above:

```python
assert list(matcher.match_all("http://example.net")) == [2]
assert list(matcher.match_all("http://example.net", include_universal=False)) == [2]
```

I think a more clear behavior would be to not return anything at all.

Maybe a name like `maybe_include_universal` is more apt? Hmmm, let me think about it more.

Reply:
always_inculde_universal=False?

Reply:
I don't think we need a short name for this argument.

Reply:
At least because it's the current behavior :)

Reply:
Ahh so I guess we won't break anything if we don't return the universal when `include_universal=False` right? since passing this parameter is quite intentional.

Reply:
If we pass the flag we want to get the universal rules if there are no domain-specific matches, because we wanted that in DUD: either you get rules specific for the domain or you get the universal ones if there are no specific ones.

Reply:
> If we pass the flag we want to get the universal rules if there are no domain-specific matches

I see. I think this is the one which affects the implementation for url-matcher.

What if in DUD, the logic would be:

1. Check if any domain-specific patterns matches a given URL.
2. If there were no results for (1), then use any universal patterns.

For both of these steps, we can have two calls for **url-matcher**. Implementing (1) is quite straightforward, except for the part in avoiding universal patterns. 

On the other hand, we could implement (2) optimally by marking any universal patterns during `matcher.add_or_update(...)`. In this way, we can have a quick reference to a list of universal patterns to return later on.

I imagine such implementations to be something like:

1. `def match_all(self, url: str, *, include_universal=True)` ⸺ but having `include_universal=False` would never return the universal IDs.
2. `def match_universal(self)` ⸺ this would only return the universal IDs and doesn't include any domain-specific ones.

What do you think?

Reply:
It would be straightforward to split the universal matchers out of `URLMatcher.matchers_by_domain` but I don't know if it's fine to change its meaning, as it's not even underscored.

Reply:
I was thinking about having no changes in `URLMatcher.matchers_by_domain`. It'd stay the same. However, we can have something like `URLMatcher.universal_matchers` which also contain the same instances of universal matchers in `URLMatcher.matchers_by_domain`.

Reply:
Yeah, I just didn't want to copy them :) I'll do it this way.